### PR TITLE
Generalize `quantize_fns` for simpler FP16 handling

### PR DESCRIPTION
**Pull request by sw**
ref: https://github.com/ggerganov/ggml/issues/286

This is hopefully in line with the goal of refactoring in the May roadmap.

In my opinion, `quantize_fns` is too specific: what matters is not whether a type is quantized, but whether it requires a row-wise conversion, either to `vec_dot_type` for the vector dot-product, or to `float` for some other operations.

By making this more generic, some specific code paths for FP16 can be eliminated, most importantly `ggml_compute_forward_mul_mat_f16_f32`.

This is not finished; CUDA and Accelerate may not work, and also the work buffer allocation may be wrong.

### CODE REVIEW

**Comment by ggerganov** [comment]
> This is hopefully in line with the goal of refactoring in the May roadmap.

Yes, this looks great

**Comment by slaren** [comment]
It would be great if `mul_mat_f32` could also be unified like this, maybe by making dummy `from/to_float` functions. I would be happy to move all the cuBLAS mul mat handling to a single mul mat function to simplify this change.

**Comment by slaren** [comment]
@sw I have moved all the cuBLAS mat mul code to a single function that should make your changes here easier, but it depends on #1233. Once that is merged, I will open a PR.

**Comment by ggerganov** [comment]
The Accelerate path works - tested the perplexity tool for quantized and F16 models

**Comment by slaren** [comment]
cuBLAS has different paths depend on the mat mul type, so this does not look right currently. #1259 should make this much easier for you since it coalesces all cuBLAS mat mul code in a function call to `ggml_cuda_mul_mat` such as:
```C++
#if defined(GGML_USE_CUBLAS)
    if (ggml_cuda_can_mul_mat(src0, src1, dst)) {
        if (params->ith == 0 && params->type == GGML_TASK_COMPUTE) {
            ggml_cuda_mul_mat(src0, src1, dst, params->wdata, params->wsize);
        }
        return;
    }
#endif
```

Incidentally, this also means that it is possible to fall back to other BLAS if cuBLAS cannot perform a mat mul, but currently that is never the case.

**Comment by goerch** [comment]
@sw : now that this is high priority and on the [roadmap](https://github.com/users/ggerganov/projects/7?pane=issue&itemId=31668809), do you want to update this patch? Otherwise I consider to look into this once https://github.com/ggerganov/ggml/pull/309 is resolved.

**Comment by slaren** [comment]
How about something like `type_traits`. In the future, it could be expanded to include other details that currently are spread around everywhere, such as the sizes or the name.

**Comment by goerch** [comment]
> I've rebased the branch onto latest master. I'm not particularly happy with the name `type_handling` though, any suggestions?

Sorry, I missed that completely. In my [PR](https://github.com/ggerganov/ggml/pull/317) against `ggml` I simply keep the original names until @ggerganov objects. But on to a more important question: I just checked and it seems parts of our changes are not covered by the existing tests. How do you verify your changes?

**Comment by ggerganov** [comment]
@goerch 
Yes, I think `type_traits` is a good suggestion.
I think we should proceed and merge this PR first as it deduplicates the forward matrix multiplication code which is nice.

@sw
Let us know when this is ready for review - it seems in a good shape to me.

**Comment by ggerganov** [comment]
Thanks - will look to merge tomorrow

File: ggml.c [hunk 1]
```diff
@@ -17002,11 +16584,13 @@ void ggml_graph_compute(struct ggml_context * ctx, struct ggml_cgraph * cgraph)
                         //printf("nr0 = %8d, nr1 = %8d, nr0*nr1 = %8d, n_tasks = %d\n", nr0, nr1, nr0*nr1, node->n_tasks);
 
                         size_t cur = 0;
+                        const enum ggml_type vec_dot_type = type_traits[node->src0->type].vec_dot_type;
 
 #if defined(GGML_USE_CUBLAS)
                         if (ggml_cuda_can_mul_mat(node->src0, node->src1, node)) {
                             node->n_tasks = 1; // TODO: this actually is doing nothing
                                                 //       the threads are still spinning
+                            cur = ggml_cuda_mul_mat_get_wsize(node->src0, node->src1, node);
```
Comment (slaren) [inline ggml.c:16593]: This line can be removed as it is no longer necessary. That should fix the cuBLAS build errors.
The function and a few others should also be removed from `ggml-cuda.h`, but that's a different issue.

**Comment by ggerganov** [comment]
I'll rebase now and run a few tests before merging

**Comment by ggerganov** [comment]
Strange - no idea why the MacOS CI is failing. It works locally both on x86 and arm

**Review by ggerganov — APPROVED** [review]
Will merge this first thing tomorrow and give it some more time for testing and hopefully resolving the CI

**Comment by mqy** [comment]
> Strange - no idea why the MacOS CI is failing. It works locally both on x86 and arm

If I remember correctly, I've experienced similar situation when bench mul_mat for f16 (the rope case) -- only macOS-cmake-latest fail with `Illegal`. I'll pull this branch and have a try.

File: ggml.c [hunk 2]
```diff
@@ -16544,37 +16189,18 @@ void ggml_graph_compute(struct ggml_context * ctx, struct ggml_cgraph * cgraph)
...
-                            } else
-#endif
-                            {
-                                const enum ggml_type type_q = quantize_fns[node->src0->type].vec_dot_type;
-                                cur = GGML_TYPE_SIZE[type_q]*ggml_nelements(node->src1)/GGML_BLCK_SIZE[type_q];
-                            }
+                        if (node->src1->type != vec_dot_type) {
+                            cur = GGML_TYPE_SIZE[vec_dot_type]*ggml_nelements(node->src1)/GGML_BLCK_SIZE[vec_dot_type];
                         } else {
                             GGML_ASSERT(false);
```
Comment (ggerganov) [inline ggml.c:16205]: We hit this assert when both `src0` and `src1` are `F32`

Comment (ggerganov) [reply ggml.c:16205]: Pushed the following fix for now: ec326d350c72afd23709a409944728a607188cc0

Probably would be best to add a unit test.
Will make an issue

**Pull request merged**